R/LearnerSurvCTree.R

```diff
@@ -0,0 +1,115 @@
+          ParamInt$new("maxsurrogate", lower = 0L, default = 0L,
+            tags = "train"),
+          ParamLgl$new("numsurrogate", default = FALSE, tags = "train"),
+          ParamInt$new("mtry", lower = 0L, special_vals = list(Inf),
+            default = Inf, tags = "train"),
+          ParamInt$new("maxdepth", lower = 0L, special_vals = list(Inf),
+            default = Inf, tags = "train"),
+          ParamLgl$new("multiway", default = FALSE, tags = "train"),
+          ParamInt$new("splittry", lower = 0L, default = 2L, tags = "train"),
+          ParamLgl$new("intersplit", default = FALSE, tags = "train"),
+          ParamLgl$new("majority", default = FALSE, tags = "train"),
+          ParamLgl$new("caseweights", default = FALSE, tags = "train"),
+          ParamUty$new("applyfun", tags = "train"),
+          ParamInt$new("cores", special_vals = list(NULL), default = NULL,
+            tags = "train"),
+          ParamLgl$new("saveinfo", default = TRUE, tags = "train"),
+          ParamLgl$new("update", default = FALSE, tags = "train"),
+          ParamFct$new("splitflavour", default = "ctree",
+            levels = c("ctree", "exhaustive"), tags = c("train", "control")),
+          ParamUty$new("offset", tags = "train"),
+          ParamUty$new("cluster", tags = "train"),
+          ParamUty$new("scores", tags = "train"),
+          ParamLgl$new("doFit", default = TRUE, tags = "train"),
+          ParamUty$new("pargs", default = mvtnorm::GenzBretz, tags = "train")
+        )
+      )
+      ps$add_dep("nresample", "testtype", CondEqual$new("MonteCarlo"))
+
+      super$initialize(
+        id            = "surv.ctree",
+        packages      = c("partykit", "coin", "distr6"),
+        feature_types = c("integer", "numeric", "factor", "ordered"),
+        predict_types = c("distr", "crank"),
+        param_set     = ps,
+        properties    = "weights"
+      )
+    }
+  ),
+
+  private = list(
+    .train = function(task) {
+      pars = self$param_set$get_values(tags = "train")
+
+      if ("weights" %in% task$properties) {
+        pars$weights = task$weights$weight
+      }
+
+      mod <- mlr3misc::invoke(partykit::ctree, formula = task$formula(),
+        data = task$data(), .args = pars)
+    },
+
+    .predict = function(task) {
+      newdata = task$data(cols = task$feature_names)
+
+      p = mlr3misc::invoke(predict, self$model, type = "prob", newdata = newdata)
+
+      # Define WeightedDiscrete distr6 distribution from the survival function
+      x = lapply(p, function(z) data.frame(x = z$time, cdf = 1 - z$surv))
+      distr = distr6::VectorDistribution$new(
+        distribution = "WeightedDiscrete",
+        params = x,
+        decorators = c("CoreStatistics", "ExoticStatistics"))
+
+      # Define crank as the mean of the survival distribution
+      crank = as.numeric(sum(x[[1]][,1] * c(x[[1]][,2][1], diff(x[[1]][,2]))))
+
+      PredictionSurv$new(task = task, crank = rep(crank, task$nrow), distr = distr)
```

Comment:
I haven't run the code but shouldn't `crank` already be a vector?

Reply:
I have to admit, I copied this from the KM Learner, which is probably not appropriate. As each element is its own survfit object, I need to loop through the list and calculate this for each element, instead of using `x[[1]]`. But is it appropriate for KM? For example if KM is stratified by categorical variable?

Reply:
We don't currently allow stratified KM, could be something to consider though. How would you imagine implementing this? Maybe including variable to stratify as a parameter?

Reply:
I would go for a "stratify" compositor or pipeop, if it doesn´t already exist as a part of vanilla mlr3.
The compositor simply chops the data up by class, and fits/predicts by class.

For this specific piece of code, this discussion isn´t too relevant though?
The "clean solution" here would be using a vectorized distribution?

Reply:
I agree with Franz. If we don't go the pipeop route then we have to create a new learner `StratifiedKaplan` which just filters the data internally, but a pipeop to do this should be relatively simple and I'm guessing already implemented

Reply:
Let me know once this part is updated and I can merge